### PR TITLE
SONARJAVA-868 RSPEC-2175 Inappropriate collection check

### PR DESCRIPTION
**Pull request by m-g-sonar**
Check on inappropriate "Collection" calls:

- Calls of methods `contains` and `remove` of any sub-type of `java.util.Collection` are now checked to control compatibility of provided argument (manage boxing and unknown type).
- Added new methods in the `ParametrizedTypeType` class to access more easily type parameters.

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+  private boolean isArgumentCompatible(Type argumentType, Type collectionParameterType) {
+    return (isSubtypeOf(argumentType, collectionParameterType) || isBoxingPossible(argumentType, collectionParameterType));
+  }
+
+  private boolean isSubtypeOf(Type argumentType, Type collectionParameterType) {
+    return argumentType.isSubtypeOf(collectionParameterType.getSymbol().getFullyQualifiedName());
+  }
+
+  private boolean isKnown(Type type) {
+    return type != null && !type.isTagged(Type.UNKNOWN);
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java]: You should not have to test for null and so inline this method.

File: java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  private boolean isBoxingPossible(Type argumentType, Type collectionParameterType) {
+    return argumentType.isPrimitive() && collectionParameterType.isPrimitiveWrapper() && boxingMatch(argumentType, collectionParameterType);
+  }
+
+  private boolean boxingMatch(Type primitiveType, Type primitiveWrapperType) {
+    String primitiveWrapperFullyQualifiedName = primitiveToPrimitiveWrapper.get(primitiveType.getSymbol().getFullyQualifiedName());
+    return primitiveWrapperFullyQualifiedName.equals(primitiveWrapperType.getSymbol().getFullyQualifiedName());
+  }
+
+  private Map<String, String> buildPrimitiveMapper() {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java]: should be a static immutable final map and more probably in Type object.

File: java-checks/src/test/files/checks/CollectionInappropriateCallsCheck.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+class A {
```
Comment (benzonico) [inline java-checks/src/test/files/checks/CollectionInappropriateCallsCheck.java]: Lacks the case of List<Number> l = Lists.<Integer>newArrayList(1);  l.contains(1);

File: java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java (hunk 1 above)
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java]: For raw types such as `List list = Lists.newArrayList();`, the test is required.

File: java-squid/src/main/java/org/sonar/java/resolve/Type.java [hunk 4]
```diff
@@ -136,6 +142,24 @@ public String toString() {
     return symbol == null ? "" : symbol.toString();
   }
 
+  @Nullable
+  public Type primitiveType() {
+    return primitiveType;
+  }
+
+  protected void setPrimitiveType(Type primitiveType) {
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/resolve/Type.java]: we truly don't need setters. The property should be package visible.

File: java-squid/src/test/java/org/sonar/java/resolve/SymbolsTest.java [hunk 5]
```diff
@@ -43,41 +43,57 @@ public void builtin_types() {
     assertThat(symbols.byteType.symbol.name).isEqualTo("byte");
```
Comment (benzonico) [inline java-squid/src/test/java/org/sonar/java/resolve/SymbolsTest.java]: What about a test on a type that is NOT a primitive.

**Comment by benzonico** [comment]
LGTM!

**Pull request closed without merging**